app/helpers/heatmap_helper.rb

```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
+      phage_clause = " AND is_phage = 1"
+    end
+
+    sort = ReportHelper.decode_sort_by(sort_by)
+
+    # fraction_subsampled was introduced 2018-03-30. For prior runs, we assume
+    # fraction_subsampled = 1.0.
+    rpm_sql = "count / (
+          (total_reads - total_ercc_reads) *
+          COALESCE(fraction_subsampled, 1.0)
+        ) * 1000 * 1000"
+
+    zscore_sql = "COALESCE(
+        GREATEST(#{ReportHelper::ZSCORE_MIN}, LEAST(#{ReportHelper::ZSCORE_MAX},
+          (#{rpm_sql} - mean) / stdev
+        )),
+        #{ReportHelper::ZSCORE_WHEN_ABSENT_FROM_BACKGROUND})"
+
+    query = "
+    SELECT
+      pipeline_run_id,
+      taxon_counts.tax_id,
+      taxon_counts.count_type,
+      taxon_counts.tax_level,
+      genus_taxid,
+      family_taxid,
+      taxon_counts.name,
+      superkingdom_taxid,
+      is_phage,
+      count               AS  r,
+      stdev,
+      mean,
+      percent_identity    AS  percentidentity,
+      alignment_length    AS  alignmentlength,
+      COALESCE(0.0 - e_value, #{ReportHelper::DEFAULT_SAMPLE_NEGLOGEVALUE}) AS neglogevalue,
+      percent_concordant  AS  percentconcordant,
+      -- First pass of ranking in SQL. Second pass in Ruby.
+      #{rpm_sql} AS rpm,
+      #{zscore_sql} AS zscore
+    FROM taxon_counts
+    -- warning!!! pipeline_runs may be missing!!!
+    LEFT OUTER JOIN pipeline_runs pr ON pipeline_run_id = pr.id
+    LEFT OUTER JOIN taxon_summaries ON
+      #{background_id.to_i}   = taxon_summaries.background_id   AND
+      taxon_counts.count_type = taxon_summaries.count_type      AND
+      taxon_counts.tax_level  = taxon_summaries.tax_level       AND
+      taxon_counts.tax_id     = taxon_summaries.tax_id
+    WHERE
+      pipeline_run_id IN (
+        SELECT MAX(id)
+        FROM pipeline_runs
+        WHERE sample_id IN (#{samples.pluck(:id).join(',')})
+        GROUP BY sample_id
+      )
+      AND genus_taxid != #{TaxonLineage::BLACKLIST_GENUS_ID}
+      AND count >= #{min_reads}
+      -- We need both types of counts for threshold filters
+      AND taxon_counts.count_type IN ('NT', 'NR')
+      #{categories_clause}
+      #{read_specificity_clause}
+      #{phage_clause}
+      AND #{zscore_sql} >= #{MINIMUM_ZSCORE_THRESHOLD}"
+
+    # This query:
+    # 1) assigns a rank to each row within a pipeline run
+    # 2) returns rows ranking <= num_results
+    # See http://www.sqlines.com/mysql/how-to/get_top_n_each_group
+    top_n_query = "
+      SELECT *
+      FROM (
+        SELECT
+          @rank := IF(@current_id = pipeline_run_id, @rank + 1, 1) AS rank,
+          @current_id := pipeline_run_id AS current_id,
+          a.*
+        FROM (
+          #{query}
+        ) a
+        ORDER BY
+          pipeline_run_id,
+          count_type = '#{sort[:count_type]}' DESC,
+          #{sort[:metric]} #{sort[:direction] == 'highest' ? 'DESC' : 'ASC'}
+      ) b
+      -- Overfetch by a factor of 4 to allow for post-SQL threshold filters
```

Comment:
Would appreciate more explanation here. Which filters are they? Why are they applied post-SQL? Is there a way to apply them in SQL? Is there a risk that the results are going to be "inexact"? Is there a way to make them exact?

Reply:
will discuss in person

Reply:
After discussion, I decided to move the logic of the threshold filters into the SQL.